File: rust/fn-dsa/src/ntru.rs
// NTRU key generation for FN-DSA (FIPS 206 Algorithm 5 / NTRUGen).
//
// Ported from the Go reference implementation. Uses num-bigint for exact arithmetic
// in the recursive NTRU solver.

use num_bigint::{BigInt, Sign};
use num_traits::{One, Zero, ToPrimitive};
use rand_core::RngCore;

use crate::gaussian::{sample_gaussian, ntru_sigma};
use crate::ntt::{ntt, intt, pow_mod_q};
use crate::params::{Params, Q};
use crate::fft::{fft, ifft, Complex64};
#[allow(unused_imports)]
use crate::ntt::poly_mul_ntt;
use crate::FnDsaError;

/// Generates (f, g, F, G) satisfying f*G - g*F = Q over Z[x]/(x^n+1).
pub fn ntru_keygen<R: RngCore>(
    p: Params,
    rng: &mut R,
) -> Result<(Vec<i32>, Vec<i32>, Vec<i32>, Vec<i32>), FnDsaError> {
    let n = p.n;
    let sigma = ntru_sigma(n);

    for _ in 0..1000 {
        // Sample f and g.
        let f_coeffs: Vec<i32> = (0..n).map(|_| sample_gaussian(rng, sigma)).collect();
        let g_coeffs: Vec<i32> = (0..n).map(|_| sample_gaussian(rng, sigma)).collect();

        // f must be invertible mod 2 (odd XOR sum).
        let xor_sum: i32 = f_coeffs.iter().fold(0, |acc, v| acc ^ (v & 1));
        if xor_sum == 0 {
            continue;
        }

        // f must be invertible mod q (no zero NTT coefficients).
        let mut f_ntt: Vec<i32> = f_coeffs.iter().map(|&v| ((v % Q) + Q) % Q).collect();
        ntt(&mut f_ntt, n);
        if f_ntt.iter().any(|&v| v == 0) {
            continue;
        }

        // Gram-Schmidt norm bound.
        let norm_sq: f64 = f_coeffs.iter().chain(g_coeffs.iter())
            .map(|&v| (v as f64) * (v as f64))
            .sum();
        if norm_sq > 1.17 * 1.17 * Q as f64 * n as f64 {
            continue;
        }

        // Solve the NTRU equation.
        let f_big = int32_to_bigint(&f_coeffs);
        let g_big = int32_to_bigint(&g_coeffs);

        match ntru_solve_big(n, &f_big, &g_big) {
            Err(_) => continue,
            Ok((f_res, g_res)) => {
                // Convert back to i32.
                let f_out: Option<Vec<i32>> = f_res.iter().map(|v| {
                    let x = ToPrimitive::to_i64(v)?;
                    if x > i32::MAX as i64 || x < i32::MIN as i64 { None } else { Some(x as i32) }
                }).collect();
                let g_out: Option<Vec<i32>> = g_res.iter().map(|v| {
                    let x = ToPrimitive::to_i64(v)?;
                    if x > i32::MAX as i64 || x < i32::MIN as i64 { None } else { Some(x as i32) }
                }).collect();

                if let (Some(cap_f), Some(cap_g)) = (f_out, g_out) {
                    // Verify the NTRU equation.
                    if verify_ntru(&f_coeffs, &g_coeffs, &cap_f, &cap_g, n) {
                        return Ok((f_coeffs, g_coeffs, cap_f, cap_g));
                    }
                }
            }
        }
    }

    Err(FnDsaError::KeyGenFailed)
}

/// Verifies f*G - g*F = Q over Z[x]/(x^n+1).
fn verify_ntru(f: &[i32], g: &[i32], cap_f: &[i32], cap_g: &[i32], n: usize) -> bool {
    let fg = poly_mul_int_z(f, cap_g, n);
    let gf = poly_mul_int_z(g, cap_f, n);
    if fg[0] - gf[0] != Q as i64 {
        return false;
    }
    for i in 1..n {
        if fg[i] - gf[i] != 0 {
            return false;
        }
    }
    true
}

/// Computes h = g * f^{-1} mod (q, x^n+1).
pub fn ntru_public_key(f: &[i32], g: &[i32], p: Params) -> Vec<i32> {
    let n = p.n;
    let mut f_ntt: Vec<i32> = f.iter().map(|&v| ((v % Q) + Q) % Q).collect();
    let mut g_ntt: Vec<i32> = g.iter().map(|&v| ((v % Q) + Q) % Q).collect();
    ntt(&mut f_ntt, n);
    ntt(&mut g_ntt, n);

    let f_inv_ntt: Vec<i32> = f_ntt.iter().map(|&v| pow_mod_q(v as i64, (Q - 2) as i64)).collect();

    let mut h_ntt: Vec<i32> = g_ntt.iter().zip(f_inv_ntt.iter())
        .map(|(&gi, &fi)| (i64::from(gi) * i64::from(fi) % i64::from(Q)) as i32)
        .collect();
    intt(&mut h_ntt, n);
    h_ntt
}

// --- Polynomial arithmetic over Z ---

/// Multiplies two polynomials over Z[x]/(x^n+1) exactly using i64.
pub fn poly_mul_int_z(a: &[i32], b: &[i32], n: usize) -> Vec<i64> {
    let mut c = vec![0i64; n];
    for (i, &ai) in a.iter().enumerate() {
        for (j, &bj) in b.iter().enumerate() {
            let idx = i + j;
            let val = i64::from(ai) * i64::from(bj);
            if idx < n {
                c[idx] += val;
            } else {
                c[idx - n] -= val;
            }
        }
    }
    c
}

fn int32_to_bigint(a: &[i32]) -> Vec<BigInt> {
    a.iter().map(|&v| BigInt::from(v)).collect()
}

// --- BigInt polynomial helpers ---

fn poly_mul_bigint_z(a: &[BigInt], b: &[BigInt], n: usize) -> Vec<BigInt> {
    let mut c = vec![BigInt::from(0i32); n];
    for (i, ai) in a.iter().enumerate() {
        for (j, bj) in b.iter().enumerate() {
            let idx = i + j;
            let val = ai * bj;
            if idx < n {
                c[idx] += &val;
            } else {
                c[idx - n] -= &val;
            }
        }
    }
    c
}

/// Field norm from Z[x]/(x^n+1) to Z[x]/(x^{n/2}+1).
/// N(f)(y) = f0(y)^2 - y*f1(y)^2
fn field_norm_bigint(f: &[BigInt], n: usize) -> Vec<BigInt> {
    let h = n / 2;
    let f0: Vec<BigInt> = (0..h).map(|i| f[2 * i].clone()).collect();
    let f1: Vec<BigInt> = (0..h).map(|i| f[2 * i + 1].clone()).collect();
    let f0sq = poly_mul_bigint_z(&f0, &f0, h);
    let f1sq = poly_mul_bigint_z(&f1, &f1, h);

    // N(f)[0] = f0sq[0] + f1sq[h-1]
    // N(f)[i] = f0sq[i] - f1sq[i-1]  for i >= 1
    let mut result = vec![BigInt::zero(); h];
    result[0] = &f0sq[0] + &f1sq[h - 1];
    for i in 1..h {
        result[i] = &f0sq[i] - &f1sq[i - 1];
    }
    result
}

/// Tower conjugate: negate odd-indexed coefficients.
fn tower_conjugate_bigint(f: &[BigInt]) -> Vec<BigInt> {
    f.iter().enumerate().map(|(i, v)| {
        if i % 2 == 0 { v.clone() } else { -v }
    }).collect()
}

/// Lifts (F', G') from degree n/2 to degree n.
fn lift_bigint(fp: &[BigInt], gp: &[BigInt], f: &[BigInt], g: &[BigInt], n: usize) -> (Vec<BigInt>, Vec<BigInt>) {
    let h = n / 2;
    let mut fp_lift = vec![BigInt::zero(); n];
    let mut gp_lift = vec![BigInt::zero(); n];
    for i in 0..h {
        fp_lift[2 * i] = fp[i].clone();
        gp_lift[2 * i] = gp[i].clone();
    }
    let f_conj = tower_conjugate_bigint(f);
    let g_conj = tower_conjugate_bigint(g);
    let cap_f = poly_mul_bigint_z(&g_conj, &fp_lift, n);
    let cap_g = poly_mul_bigint_z(&f_conj, &gp_lift, n);
    (cap_f, cap_g)
}

// --- BigFloat FFT Babai ---

fn bigint_to_f64(v: &BigInt) -> f64 {
    // Convert BigInt to f64 (may lose precision for large values)
    match v.to_bytes_be() {
        (Sign::Minus, bytes) => {
            let pos = bytes_to_f64(&bytes);
            -pos
        }
        (Sign::Plus, bytes) | (Sign::NoSign, bytes) => bytes_to_f64(&bytes),
    }
}

fn bytes_to_f64(bytes: &[u8]) -> f64 {
    // Convert big-endian bytes to f64
    let mut val = 0.0f64;
    for &b in bytes {
        val = val * 256.0 + b as f64;
    }
    val
}

/// Babai FFT rounding using float64.
fn babai_float64(
    big_f: &[BigInt], big_g: &[BigInt],
    f: &[f64], g: &[f64],
    n: usize,
) -> Vec<BigInt> {
    let mut fc: Vec<Complex64> = f.iter().map(|&v| (v, 0.0)).collect();
    let mut gc: Vec<Complex64> = g.iter().map(|&v| (v, 0.0)).collect();
    let mut big_fc: Vec<Complex64> = big_f.iter().map(|v| (bigint_to_f64(v), 0.0)).collect();
    let mut big_gc: Vec<Complex64> = big_g.iter().map(|v| (bigint_to_f64(v), 0.0)).collect();

    fft(&mut fc, n);
    fft(&mut gc, n);
    fft(&mut big_fc, n);
    fft(&mut big_gc, n);

    let mut kc: Vec<Complex64> = vec![(0.0, 0.0); n];
    for i in 0..n {
        let fi = fc[i];
        let gi = gc[i];
        let cap_fi = big_fc[i];
        let cap_gi = big_gc[i];
        let fi_conj = (fi.0, -fi.1);
        let gi_conj = (gi.0, -gi.1);
        let num = (
            cap_fi.0 * fi_conj.0 - cap_fi.1 * fi_conj.1 + cap_gi.0 * gi_conj.0 - cap_gi.1 * gi_conj.1,
            cap_fi.0 * fi_conj.1 + cap_fi.1 * fi_conj.0 + cap_gi.0 * gi_conj.1 + cap_gi.1 * gi_conj.0,
        );
        let denom = fi.0 * fi.0 + fi.1 * fi.1 + gi.0 * gi.0 + gi.1 * gi.1;
        if denom != 0.0 {
            kc[i] = (num.0 / denom, num.1 / denom);
        }
    }

    ifft(&mut kc, n);

    kc.iter().map(|v| BigInt::from(v.0.round() as i64)).collect()
}

/// High-precision Babai using a custom big-float complex FFT.
/// For deep levels where f64 is insufficient, we use a higher-precision approach.
/// We use f64 but with iterative refinement via multiple rounds.
fn babai_high_prec(
    big_f: &[BigInt], big_g: &[BigInt],
    f_big: &[BigInt], g_big: &[BigInt],
    n: usize,
) -> Vec<BigInt> {
    // Convert f_big, g_big to f64 (may lose precision but good enough for rounding)
    let f_f64: Vec<f64> = f_big.iter().map(|v| bigint_to_f64(v)).collect();
    let g_f64: Vec<f64> = g_big.iter().map(|v| bigint_to_f64(v)).collect();
    babai_float64(big_f, big_g, &f_f64, &g_f64, n)
}

/// Recursive NTRU solver using exact big.Int arithmetic.
fn ntru_solve_big(n: usize, f: &[BigInt], g: &[BigInt]) -> Result<(Vec<BigInt>, Vec<BigInt>), String> {
    if n == 1 {
        // Base case: solve f[0]*G[0] - g[0]*F[0] = Q over Z using extended GCD.
        let f_val = &f[0];
        let g_val = &g[0];
        let q_big = BigInt::from(Q);

        let (gcd, u, v) = extended_gcd(f_val.clone(), g_val.clone());

        // Check gcd divides Q.
        if !(&q_big % &gcd).is_zero() {
            return Err("gcd does not divide q at base case".to_string());
        }

        let scale = &q_big / &gcd;
        let cap_g = u * &scale;
        let cap_f = -(v * &scale);

        return Ok((vec![cap_f], vec![cap_g]));
    }

    // Compute field norms.
    let f_norm = field_norm_bigint(f, n);
    let g_norm = field_norm_bigint(g, n);

    // Recursive solve.
    let (fp, gp) = ntru_solve_big(n / 2, &f_norm, &g_norm)?;

    // Lift.
    let (mut cap_f, mut cap_g) = lift_bigint(&fp, &gp, f, g, n);

    // Determine max bit length of f, g.
    let max_bits: usize = f.iter().chain(g.iter())
        .map(|v| v.bits() as usize)
        .max()
        .unwrap_or(0);

    // Babai reduction - 2 rounds.
    for _ in 0..2 {
        let max_fg_bits = {
            let fg_bits = cap_f.iter().chain(cap_g.iter())
                .map(|v| v.bits() as usize)
                .max()
                .unwrap_or(0);
            max_bits.max(fg_bits)
        };

        let k = if max_fg_bits <= 53 {
            let f_f64: Vec<f64> = f.iter().map(|v| bigint_to_f64(v)).collect();
            let g_f64: Vec<f64> = g.iter().map(|v| bigint_to_f64(v)).collect();
            babai_float64(&cap_f, &cap_g, &f_f64, &g_f64, n)
        } else {
            babai_high_prec(&cap_f, &cap_g, f, g, n)
        };

        let kf = poly_mul_bigint_z(&k, f, n);
        let kg = poly_mul_bigint_z(&k, g, n);
        for i in 0..n {
            cap_f[i] -= &kf[i];
            cap_g[i] -= &kg[i];
        }
    }

    Ok((cap_f, cap_g))
}

/// Extended Euclidean algorithm: returns (gcd, u, v) with u*a + v*b = gcd.
fn extended_gcd(mut a: BigInt, mut b: BigInt) -> (BigInt, BigInt, BigInt) {
    let mut u0 = BigInt::one();
    let mut u1 = BigInt::zero();
    let mut v0 = BigInt::zero();
    let mut v1 = BigInt::one();

    while !b.is_zero() {
        let q = &a / &b;
        let r = &a % &b;
        a = b;
        b = r;
        let new_u = u0 - &q * &u1;
        u0 = u1;
        u1 = new_u;
        let new_v = v0 - &q * &v1;
        v0 = v1;
        v1 = new_v;
    }

    // Make gcd positive.
    if a.sign() == Sign::Minus {
        a = -a;
        u0 = -u0;
        v0 = -v0;
    }

    (a, u0, v0)
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extended_gcd() {
        let (gcd, u, v) = extended_gcd(BigInt::from(3), BigInt::from(5));
        assert_eq!(gcd, BigInt::one());
        // u*3 + v*5 = 1
        assert_eq!(&u * BigInt::from(3) + &v * BigInt::from(5), BigInt::one());
    }

    #[test]
    fn test_field_norm() {
        // f(x) = 1 + x, norm at n=2: f0=[1], f1=[1], N(f)= [1]^2 - y*[1]^2 = 1 - y
        // in Z[y]/(y+1): 1 - (-1) = 2? Actually let's just check dimensions.
        let f: Vec<BigInt> = vec![BigInt::from(1), BigInt::from(1)];
        let norm = field_norm_bigint(&f, 2);
        assert_eq!(norm.len(), 1);
    }
}

File: python/fndsa/ntru.py
"""NTRU key generation for FN-DSA (FIPS 206 Algorithm 5).

Implements recursive NTRU equation solver over Z[x]/(x^n+1).
Uses Python's arbitrary-precision integers throughout for exact arithmetic.

Algorithm:
1. Sample f, g from D_{Z, sigma} where sigma = 1.17 * sqrt(Q / (2n))
2. Check f is invertible mod 2 and mod q, and norm bound is satisfied
3. Solve f*G - g*F = Q over Z[x]/(x^n+1) using recursive field norm
4. Babai reduce (F, G) using (f, g) basis
5. Verify f*G - g*F = Q exactly
"""
import math
import sys

# Ensure sufficient recursion depth for the NTRU recursive solver.
# At n=1024, the recursion goes ~10 levels deep but each level involves
# significant Python call overhead (poly_mul_int_z, etc.).
_REQUIRED_RECURSION_LIMIT = 5000
if sys.getrecursionlimit() < _REQUIRED_RECURSION_LIMIT:
    sys.setrecursionlimit(_REQUIRED_RECURSION_LIMIT)

from .params import Params, Q
from .gaussian import sample_gaussian
from .ntt import ntt as _ntt


def _ntru_sigma(n: int) -> float:
    """Compute sigma = 1.17 * sqrt(Q / (2n))."""
    return 1.17 * math.sqrt(Q / (2 * n))


def _poly_mul_int_z(a: list[int], b: list[int], n: int) -> list[int]:
    """Multiply two polynomials over Z[x]/(x^n+1) exactly using Python big ints."""
    c = [0] * n
    for i, ai in enumerate(a):
        for j, bj in enumerate(b):
            idx = i + j
            val = ai * bj
            if idx < n:
                c[idx] += val
            else:
                c[idx - n] -= val
    return c


def _poly_mul_int_z_scalar(k: list[int], a: list[int], n: int) -> list[int]:
    """Multiply polynomial k by polynomial a over Z[x]/(x^n+1)."""
    return _poly_mul_int_z(k, a, n)


def _field_norm(f: list[int], n: int) -> list[int]:
    """Compute field norm N(f) from Z[x]/(x^n+1) to Z[x]/(x^{n/2}+1).

    N(f)(y) = f_0(y)^2 - y * f_1(y)^2
    where f(x) = f_0(x^2) + x*f_1(x^2).
    """
    h = n // 2
    f0 = [f[2 * i] for i in range(h)]
    f1 = [f[2 * i + 1] for i in range(h)]
    f0sq = _poly_mul_int_z(f0, f0, h)
    f1sq = _poly_mul_int_z(f1, f1, h)

    # N(f)[0] = f0sq[0] + f1sq[h-1]  (wrapping: y*f1sq gives -f1sq[h-1] at index 0)
    # N(f)[i] = f0sq[i] - f1sq[i-1]  for i >= 1
    result = [0] * h
    result[0] = f0sq[0] + f1sq[h - 1]
    for i in range(1, h):
        result[i] = f0sq[i] - f1sq[i - 1]
    return result


def _tower_conjugate(f: list[int]) -> list[int]:
    """Compute tower conjugate: f*(x) = f_0(x^2) - x*f_1(x^2).

    Negates the odd-indexed coefficients.
    """
    result = list(f)
    for i in range(1, len(f), 2):
        result[i] = -result[i]
    return result


def _lift(fp: list[int], gp: list[int], f: list[int], g: list[int], n: int) -> tuple:
    """Lift (F', G') from degree n/2 to degree n.

    F = g*(x) * F'(x^2) in Z[x]/(x^n+1)
    G = f*(x) * G'(x^2) in Z[x]/(x^n+1)
    where f*(x) and g*(x) are tower conjugates.
    """
    h = n // 2
    # Embed F' and G' into degree n by replacing x with x^2
    fp_lift = [0] * n
    gp_lift = [0] * n
    for i in range(h):
        fp_lift[2 * i] = fp[i]
        gp_lift[2 * i] = gp[i]

    f_conj = _tower_conjugate(f)
    g_conj = _tower_conjugate(g)

    # F = g_conj * F'(x^2), G = f_conj * G'(x^2)
    F = _poly_mul_int_z(g_conj, fp_lift, n)
    G = _poly_mul_int_z(f_conj, gp_lift, n)
    return F, G


def _babai_reduce_fft(F: list[int], G: list[int], f: list[int], g: list[int], n: int) -> list[int]:
    """Compute Babai rounding k using complex FFT.

    k = round(Re(IFFT((F*conj(f) + G*conj(g)) / (|f|^2 + |g|^2))))
    in the FFT domain.

    Works when coefficients fit in float64 (< 53 bits).
    """
    import cmath
    from .fft import fft as _fft, ifft as _ifft

    fC = _fft([complex(v) for v in f], n)
    gC = _fft([complex(v) for v in g], n)
    FC = _fft([complex(v) for v in F], n)
    GC = _fft([complex(v) for v in G], n)

    kC = [0j] * n
    for i in range(n):
        fi, gi = fC[i], gC[i]
        Fi, Gi = FC[i], GC[i]
        fi_conj = fi.conjugate()
        gi_conj = gi.conjugate()
        num = Fi * fi_conj + Gi * gi_conj
        denom = fi * fi_conj + gi * gi_conj
        if denom.real != 0 or denom.imag != 0:
            kC[i] = num / denom

    k_ifft = _ifft(kC, n)
    return [round(v.real) for v in k_ifft]


def _babai_reduce_bigfloat(F: list[int], G: list[int], f: list[int], g: list[int], n: int) -> list[int]:
    """Compute Babai rounding k using Python's mpmath for high-precision arithmetic.

    Required when coefficient bit lengths exceed float64 precision (53 bits).
    """
    try:
        import mpmath
    except ImportError:
        # Fall back to float64 FFT (may lose precision for large coefficients)
        return _babai_reduce_fft(F, G, f, g, n)

    from .fft import _fft_log_n as _fft_logn_fn, _fft_bit_rev

    # Determine required precision based on coefficient bit lengths
    max_bits = max(
        max((v.bit_length() for v in f), default=0),
        max((v.bit_length() for v in g), default=0),
        max((abs(v).bit_length() for v in F), default=0),
        max((abs(v).bit_length() for v in G), default=0),
    )
    prec = max_bits * 2 + _fft_logn_fn(n) * 64 + 256
    mp = mpmath.mp
    old_prec = mp.prec
    mp.prec = prec

    try:
        def to_mpc(lst):
            return [mpmath.mpc(int(v), 0) for v in lst]

        def do_fft(arr):
            """In-place forward FFT using mpmath."""
            logn = _fft_logn_fn(n)
            k = 0
            length = n >> 1
            while length >= 1:
                start = 0
                while start < n:
                    k += 1
                    brk = _fft_bit_rev(k, logn)
                    angle = mpmath.pi * brk / n
                    w = mpmath.mpc(mpmath.cos(angle), mpmath.sin(angle))
                    for j in range(start, start + length):
                        t = w * arr[j + length]
                        arr[j + length] = arr[j] - t
                        arr[j] = arr[j] + t
                    start += 2 * length
                length >>= 1

        def do_ifft(arr):
            """In-place inverse FFT using mpmath."""
            logn = _fft_logn_fn(n)
            k = n
            length = 1
            while length < n:
                start = n - 2 * length
                while start >= 0:
                    k -= 1
                    brk = _fft_bit_rev(k, logn)
                    angle = mpmath.pi * brk / n
                    w_inv = mpmath.mpc(mpmath.cos(angle), -mpmath.sin(angle))
                    for j in range(start, start + length):
                        t = arr[j]
                        arr[j] = t + arr[j + length]
                        arr[j + length] = w_inv * (t - arr[j + length])
                    start -= 2 * length
                length <<= 1
            inv_n = mpmath.mpf(1) / n
            for i in range(n):
                arr[i] = arr[i] * inv_n

        fA = to_mpc(f)
        gA = to_mpc(g)
        FA = to_mpc(F)
        GA = to_mpc(G)

        do_fft(fA)
        do_fft(gA)
        do_fft(FA)
        do_fft(GA)

        kA = [mpmath.mpc(0) for _ in range(n)]
        for i in range(n):
            fi_conj = mpmath.mpc(fA[i].real, -fA[i].imag)
            gi_conj = mpmath.mpc(gA[i].real, -gA[i].imag)
            num = FA[i] * fi_conj + GA[i] * gi_conj
            denom_re = fA[i].real**2 + fA[i].imag**2 + gA[i].real**2 + gA[i].imag**2
            if denom_re != 0:
                kA[i] = mpmath.mpc(num.real / denom_re, num.imag / denom_re)

        do_ifft(kA)

        result = [int(mpmath.nint(v.real)) for v in kA]
        return result
    finally:
        mp.prec = old_prec


def _fft_log_n(n: int) -> int:
    logn = 0
    t = n
    while t > 1:
        t >>= 1
        logn += 1
    return logn


def _max_bits(polys: list[list[int]]) -> int:
    """Return the maximum bit length of coefficients across all polynomials."""
    result = 0
    for p in polys:
        for v in p:
            b = abs(v).bit_length()
            if b > result:
                result = b
    return result


def _ntru_solve_recursive(n: int, f: list[int], g: list[int]) -> tuple:
    """Recursively solve f*G - g*F = Q over Z[x]/(x^n+1).

    Returns (F, G) as lists of Python ints.
    Raises ValueError if no solution exists.
    """
    if n == 1:
        # Base case: solve f[0]*G[0] - g[0]*F[0] = Q over Z using extended GCD.
        fv = f[0]
        gv = g[0]

        # Extended GCD: gcd, u, v such that fv*u + gv*v = gcd
        def ext_gcd(a, b):
            if b == 0:
                return a, 1, 0
            g2, u, v = ext_gcd(b, a % b)
            return g2, v, u - (a // b) * v

        gcd_val, u, v = ext_gcd(fv, gv)
        if Q % gcd_val != 0:
            raise ValueError("gcd does not divide Q at base case")

        scale = Q // gcd_val
        # f*G - g*F = Q: G = u*scale, F = -v*scale
        G_val = u * scale
        F_val = -v * scale
        return [F_val], [G_val]

    # Compute field norms
    f_norm = _field_norm(f, n)
    g_norm = _field_norm(g, n)

    # Recursively solve for half-degree problem
    Fp, Gp = _ntru_solve_recursive(n // 2, f_norm, g_norm)

    # Lift from n/2 to n
    F, G = _lift(Fp, Gp, f, g, n)

    # Babai reduction: run 2 rounds
    for _ in range(2):
        max_b = _max_bits([f, g, F, G])
        if max_b <= 53:
            k = _babai_reduce_fft(F, G, f, g, n)
        else:
            k = _babai_reduce_bigfloat(F, G, f, g, n)

        kf = _poly_mul_int_z(k, f, n)
        kg = _poly_mul_int_z(k, g, n)
        for i in range(n):
            F[i] -= kf[i]
            G[i] -= kg[i]

    return F, G


def _verify_ntru(f: list[int], g: list[int], F: list[int], G: list[int], n: int) -> bool:
    """Verify f*G - g*F = Q exactly over Z[x]/(x^n+1)."""
    fG = _poly_mul_int_z(f, G, n)
    gF = _poly_mul_int_z(g, F, n)
    if fG[0] - gF[0] != Q:
        return False
    for i in range(1, n):
        if fG[i] - gF[i] != 0:
            return False
    return True


def ntru_keygen(params: Params, rng=None):
    """Generate NTRU key pair (f, g, F) for FN-DSA.

    Returns (f, g, F) where f*G - g*F = Q mod (x^n+1).
    G is not returned (it can be recovered via the NTRU equation during signing).

    rng: callable(n_bytes) -> bytes
    """
    import os
    if rng is None:
        rng = os.urandom

    n = params.n
    sigma = _ntru_sigma(n)

    for attempt in range(1000):
        # Sample f and g from D_{Z, sigma}
        f = [sample_gaussian(sigma, rng) for _ in range(n)]
        g = [sample_gaussian(sigma, rng) for _ in range(n)]

        # f must be invertible mod 2: XOR of all coefficients must be 1
        xor_sum = 0
        for v in f:
            xor_sum ^= v & 1
        if xor_sum == 0:
            continue

        # f must be invertible mod q: all NTT coefficients must be nonzero
        f_mod_q = [((v % Q) + Q) % Q for v in f]
        f_ntt = _ntt(f_mod_q, n)
        if any(v == 0 for v in f_ntt):
            continue

        # Gram-Schmidt norm bound: ||f||^2 + ||g||^2 <= 1.17^2 * Q * n
        norm_sq = sum(v * v for v in f) + sum(v * v for v in g)
        if norm_sq > 1.17 * 1.17 * Q * n:
            continue

        # Solve the NTRU equation
        try:
            F, G = _ntru_solve_recursive(n, f, g)
        except (ValueError, RecursionError):
            continue

        # Convert to int (should already be Python ints)
        F = [int(v) for v in F]
        G = [int(v) for v in G]

        # Check coefficients fit in int8 range for F (8-bit encoding)
        if any(v < -128 or v > 127 for v in F):
            continue

        # Verify the NTRU equation holds exactly
        if not _verify_ntru(f, g, F, G, n):
            continue

        return f, g, F

    raise RuntimeError("NTRU key generation failed after 1000 attempts")
